### PR TITLE
Feat: recurrent ippo vmapped agents

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
First fully end to end JAX recurrent IPPO Agent that is able to train on Rware tiny-4ag. 

## Note
This is joint work done with @RuanJohn  and @DriesSmit.

### CODE REVIEW

File: jax_distribution/types.py [hunk 1]
```diff
@@ -43,6 +43,18 @@ class LearnerState(NamedTuple):
...
+class RNNLearnerState(NamedTuple):
+    """State of the `Learner` for recurrent architectures."""
+
+    params: FrozenDict
+    opt_state: OptState
+    key: chex.PRNGKey
+    env_state: LogEnvState
+    timestep: TimeStep
+    dones: chex.Array
+    hstate: chex.Array
```
Comment (RuanJohn) [inline jax_distribution/types.py:55]: Something that could be worth look at is having this be something like 
`hstates: Union[chex.Array, Tuple[chex.Array, chex.Array]]`
This could enable use to reuse the same `NamedTuple` for rec ippo and rec mappo. It might be a bit confusing in the rec ippo case to the have a plural word here when we only have a single hidden state though. Keen to hear your and @sash-a's thoughts on this.

File: jax_distribution/types.py [hunk 2]
```diff
@@ -53,6 +65,18 @@ class EvalState(NamedTuple):
...
 
 
+class RNNEvalState(NamedTuple):
+    """State of the evaluator for recurrent architectures."""
+
+    key: chex.PRNGKey
+    env_state: State
+    timestep: TimeStep
+    dones: chex.Array
+    init_hstate: chex.Array
```
Comment (RuanJohn) [inline jax_distribution/types.py:75]: Similarly here to above.

File: jax_distribution/types.py (hunk 1 above)
Comment (sash-a) [inline jax_distribution/types.py:55]: Can we make `LearnerState` a `chex.dataclass` and then 
```suggestion
@chex.dataclass
class RNNLearnerState(LearnerState):
    dones: chex.Array
    hstate: chex.Array
```

File: jax_distribution/types.py [hunk 3]
```diff
@@ -53,6 +65,18 @@ class EvalState(NamedTuple):
     return_: chex.Numeric = None
 
 
+class RNNEvalState(NamedTuple):
```
Comment (sash-a) [inline jax_distribution/types.py:68]: Also same as above

File: jax_distribution/evaluator.py [hunk 4]
```diff
@@ -124,35 +125,185 @@ def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOut
...
+        evaluator = get_ff_evaluator_fn(
+            eval_env,
+            vmapped_eval_network_apply_fn,
+            config,
+        )
+        absolute_metric_evaluator = get_ff_evaluator_fn(
+            eval_env,
+            vmapped_eval_network_apply_fn,
+            config,
+            10,
```
Comment (sash-a) [inline jax_distribution/evaluator.py:305]:
```suggestion
            10,  # todo: add this to the config
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Couple quite minor things, but otherwise looking great

File: jax_distribution/evaluator.py [hunk 5]
```diff
@@ -124,35 +125,185 @@ def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOut
...
+    return evaluator_fn
+
+
 def evaluator_setup(
     eval_env: Environment,
     rng_e: chex.PRNGKey,
     network: Any,
     params: FrozenDict,
     config: Dict,
+    rnn_net: bool = False,
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jax_distribution/evaluator.py:266]: Maybe even something like `recurrent_policy` ?

File: jax_distribution/evaluator.py [hunk 6]
```diff
@@ -124,35 +125,185 @@ def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOut
...
+            vmapped_eval_network_apply_fn,
+            config,
+            ScannedRNN,
+        )
+        absolute_metric_evaluator = get_rnn_evaluator_fn(
+            eval_env,
+            vmapped_eval_network_apply_fn,
+            config,
+            ScannedRNN,
+            10,
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jax_distribution/evaluator.py:289]: The absolute metric is suggested to be 10. So I don't even think we need it in the config. 
Maybe just a comment to say that this is 10 because of the default absolute metric requirements.

File: jax_distribution/evaluator.py (hunk 4 above)
Comment (RuanJohn) [reply jax_distribution/evaluator.py:305]: Same as above.

File: jax_distribution/systems/rec_ippo_rware.py [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+                lambda x: jnp.swapaxes(
+                    jnp.reshape(
+                        x,
+                        [x.shape[0], config["num_minibatches"], -1] + list(x.shape[2:]),
+                    ),
+                    1,
+                    0,
+                ),
+                shuffled_batch,
+            )
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jax_distribution/systems/rec_ippo_rware.py:374]: Totally agree on this one. 
Maybe just `shaped_batch` -> `reshaped_batch`.

File: jax_distribution/evaluator.py (hunk 4 above)
Comment (OmaymaMahjoub) [reply jax_distribution/evaluator.py:305]: should we update it now better than keep it to do?

File: jax_distribution/evaluator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sash-a) [reply jax_distribution/evaluator.py:266]: I'd really like the `use` in there, just because it makes it clear that it's a `bool`, so `use_recurrent_policy`?

[the remark this answers is not in the document]
Comment (RuanJohn) [reply jax_distribution/evaluator.py:266]: Makes sense, I think the original `use_recurrent_net` is the best option here. Since we could have recurrent policies and critics.

**Review by sash-a — APPROVED** [review]
Looks good to me :smile:

File: jax_distribution/systems/rec_ippo_rware.py [hunk 8]
```diff
@@ -361,17 +352,14 @@ def _loss_fn(
...
             )
-            minibatches = jax.tree_util.tree_map(
-                lambda x: jnp.swapaxes(
-                    jnp.reshape(
-                        x,
-                        [x.shape[0], config["num_minibatches"], -1] + list(x.shape[2:]),
-                    ),
-                    1,
-                    0,
+            reshaped_batch = jax.tree_util.tree_map(
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jax_distribution/systems/rec_ippo_rware.py:355]: Sounds good!
